draft-ietf-ppm-dap.md

```diff
@@ -1547,6 +1552,66 @@ the Leader to decrypt its input share. The Leader also cannot validate the
 Helper's extensions because it cannot decrypt the Helper's input share.
 Mandatory validation of extensions will occur during aggregation.)
 
+### Bulk upload {#bulk-upload}
+
+Clients or intermediates with multiple reports in hand can use bulk upload endpoint
+and send an HTTP POST to `{leader}/tasks/{task-id}/reports/bulk` with media type
+"application/dap-bulk-report". A bulk request is equivalent to multiple separate
+submissions and is strictly a performance optimization.
+
+The "application/dap-bulk-report" format contains a prologue that encodes extensions that are
+common to all reports. The header is followed by any number of reports, from which those
+shared fields are removed.
+
+~~~ tls-presentation
+struct {
+Extension common_extensions<0..2^16-1>;
+Report report[REPORT_COUNT];
+} BulkReport;
+~~~
+{: #syn-bulk-report title="Bulk Report Format"}
+
+The use of `REPORT_COUNT` in {{syn-bulk-report}} is a small abuse of the TLS syntax
+to signify that any number of reports are included. This "value" could be any
+positive integer. Unlike a variable-length field, as denoted with `&lt;` and `>`,
+this format does not require that the size of all included reports
+be known before constructing a request.
+
+The `common_extensions` field contains common extensions that are added to the set of
+public report extensions in each report that follows. Reports that include values for
+any common extension override the value in the common extension.
+
+Each individual report from the bulk upload request will be processed as described in {{upload-request}}.
+
+### Partial failures
+
+Bulk uploads take longer time to complete and have higher chance to fail, leaving clients
+uncertain about the status of their uploads. Replaying the entire request again
+wastes both server and client resources and network bandwidth.
+
+To provide  more robust failure handling, reports in the upload MUST be implicitly
+enumerated by the Server using 0-based indexing. Monotonically increasing report index
+is used to communicate the upload status to the Client.
+
+Any report in the upload may be rejected for reasons listed in ({{upload-request}}).
+
+Leader SHOULD provide a means to Clients to obtain status of the upload.
+
+~~~ tls-presentation
+struct {
+/* The index of the latest report processed so far by the server */
+uint32 latest_index,
+/* The list of indices of reports rejected on the server side */
+uint32 rejected<..>,
+} BulkReportStatus;
+~~~
+
+Server MAY compress large sequences of rejected reports into ranges, to avoid payload blowout.
```

Comment:
This should be fully specified, since "compress ... into ranges" can have multiple interpretations.